Working code, small tests

Sword-Smith opened the pull request:
This is how the `RootFromAuthStruct` implementation could look in Rust. Code works for these two very small test cases.

Sword-Smith committed:
feat(mmr_auth_struct): `root_from_authentication_struct` algorithm

Add a function to calculate a Merkle root from an authentication struct
in a way that conductive to ZK calculations. And add a function to
generate the relevant witness data for this calculation. The
authentication structure is a list of digests that is identical to what
can be found in `authentication_structure` field in the
`MerkleTreeInclusionProof`. With appropriate witness structure, this
authentication structure can be verified faster than verifying each
individual leaf in an MMR, we hope. This
`root_from_authentication_struct` function now needs to be implemented
in `tasm-lib` to see if this actually gives shorter programs in
Triton-VM.

The purpose of this cryptography is to speed up the program that
verifies that no double-spends are occurring.

See #228.

aszepieniec commented on twenty-first/src/util_types/merkle_tree.rs:
```diff
@@ -1033,4 +1042,166 @@ pub mod merkle_tree_test {
+        assert_eq!(auth_struct.len(), auth_struct_indices.len());
+        let (alpha, gamma, delta) = {
+            let mut sponge = Tip5::init();
+            sponge.pad_and_absorb_all(&leafs.encode());
+            sponge.pad_and_absorb_all(&auth_struct.encode());
```
This is a non-standard mode of use of the sponge construction. While I cannot find an attack, I am concerned that a vulnerability might hide here. If we definitely want this construction (for instance, because it is more efficient to compute in Triton VM) then we need an iron-clad security argument.

The alternative would be something like `sponge.pad_and_absorb_all(&(leafs, auth_struct).encode())`.

Sword-Smith replied:
Your suggestion probably requires allocation in Triton VM. It's preferable if we can avoid that as much as possible.

aszepieniec commented on twenty-first/src/util_types/merkle_tree.rs:
```diff
@@ -1033,4 +1042,166 @@ pub mod merkle_tree_test {
+        for i in 0..auth_struct.len() {
+            let leaf_index_hi = bfe!(auth_struct_indices[i] >> 32);
+            let leaf_index_lo = bfe!(auth_struct_indices[i] & u32::MAX as u64);
+            let leaf_index_xfe =
+                XFieldElement::new([leaf_index_hi, leaf_index_lo, BFieldElement::zero()]);
```
Why are you putting the more significant limb (`hi`) on the least significant coefficient, but nothing on the most significant coefficient? It's counter-intuitive.

aszepieniec commented on twenty-first/src/util_types/merkle_tree.rs:
```diff
@@ -1033,4 +1042,166 @@ pub mod merkle_tree_test {
+            p *= fact1.inverse() * fact2.inverse() * fact_parent;
+        }
+
+        let x_as_xfe = XFieldElement::new([
```
A better name might be `index_1_as_xfe`.

aszepieniec replied:
Or `root_index_as_xfe`.

aszepieniec reviewed (commented):
The algorithm looks correct to me. I left a couple of nits inline.

Do you want to merge this into `twenty-first`? The repository doesn't have any other functions that take nondeterminism. You can make it fit though by planting the relevant function on a data structure, similarly to what was done with `MmrSuccessorProof`. In this case the right struct might be `AuthStructIntegrityProof` or something like that, and the function `verify(&self, leafs : &[Digest], leaf_indices: &[u64], auth_struct: &[Digest], root: Digest) -> bool`.

I intentionally left the indices of the auth path elements from this function signature because they can be inferred from the leaf indices. They are not part of the current authentication structure. Come to think of it, rather than computing the auth struct indices, it probably makes more sense to divine them in and then verify them.

Since the indices are `u64`s, this method and/or struct probably belongs in `mmr/` rather than `merkle_tree.rs`.

Sword-Smith commented:
Moved code to `mmr` per your suggestion. auth-struct indices are now divined in, not part of the public data. Also added many tests.

Sword-Smith commented:
So we have a nice solution that works for Merkle trees. How do we translate this to something that works for Merkle mountain ranges? Do we just include an authentication structure for each peak that the list of leafs are contained in?

aszepieniec commented:
> So we have a nice solution that works for Merkle trees. How do we translate this to something that works for Merkle mountain ranges? Do we just include an authentication structure for each peak that the list of leafs are contained in?

I don't see how things could be done more efficiently, seen as how the various trees in an MMR are completely uncorrelated.

Sword-Smith committed:
refactor(mmr_authentication_struct): Encapsulate code for nd-sibling indices

Sword-Smith committed:
docs(mmr_authentication_struct): Add note about better way of finding nd-sibling indices

Sword-Smith committed:
feat(mmr_authentication_struct): Derive from MMR & MMR MPs

Add a function to derive an MMR authentication structure from an MMR
accumulator and a list of indexed MMR membership proofs.

Sword-Smith commented:
Now this data structure and associated data can be derived from two sources: Either from an (MMR, membership-proof list) pair, or from a Merkle tree and a list of indices into this tree.

Sword-Smith committed:
refactor(mmr_authnetication_struct): Cleanup MMR -> auth struct code

Sword-Smith committed:
mmr_authentication_struct: Simplify derivation of nd-sibling indices

The work to find these indices was already done in a helper function. No
need to derive them again inside of the constructor-functions.

Sword-Smith committed:
mmr_authentication_struct: Build multiple witnesses from one MMR

Add a struct for encapulating the Merkle authentication struct
authenticity witness and one for encapsulating the auth struct, its
withness, and the data, which are the indexed leaves.

Sword-Smith committed:
mmr_authentication_proof: Apply Alan's suggested name for this witness/proof

Sword-Smith committed:
mmr_authentication_struct: Add for empty MMR / no opened leafs

Sword-Smith commented:
I'm happy with the state of this code now. Now the only question is just whether a) we actually want to use this data structure, and b) if we do want to use it, whether it belongs in this repo.

Oh, and c) whether the interface of `root_from_authentication_struct` is conducive to a short execution trace in Triton VM.

Sword-Smith committed:
test(mmr_authentication_struct): Verify panic on missing ND

Sword-Smith committed:
style(mmr_authentication_struct): Fix some bad variable names

jan-ferdinand commented:
@Sword-Smith, is this PR still relevant?
